src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkBatchInsertRecord.java

```diff
@@ -0,0 +1,322 @@
+import java.util.Set;
+
+/**
+ * A simple implementation of the ISQLServerBulkRecord interface that can be used to read in the basic Java data types from an ArrayList of Parameters
+ * that were provided by pstmt/cstmt.
+ */
+public class SQLServerBulkBatchInsertRecord extends SQLServerBulkCommon implements ISQLServerBulkRecord, java.lang.AutoCloseable {
+
+    private ArrayList<Parameter[]> batchParam;
+    private int batchParamIndex = -1;
+    private ArrayList<String> columnList;
+    private ArrayList<String> valueList;
+
+    public SQLServerBulkBatchInsertRecord(ArrayList<Parameter[]> batchParam,
+            ArrayList<String> columnList,
+            ArrayList<String> valueList,
+            String encoding) throws SQLServerException {
+        loggerClassName = "com.microsoft.sqlserver.jdbc.SQLServerBulkBatchInsertRecord";
+        loggerExternal = java.util.logging.Logger.getLogger(loggerClassName);
+        loggerExternal.entering(loggerClassName, "SQLServerBulkBatchInsertRecord", new Object[] {batchParam, encoding});
+
+        if (null == batchParam) {
+            throwInvalidArgument("batchParam");
+        }
+
+        if (null == valueList) {
+            throwInvalidArgument("valueList");
+        }
+
+        this.batchParam = batchParam;
+        this.columnList = columnList;
+        this.valueList = valueList;
+        columnMetadata = new HashMap<>();
+
+        loggerExternal.exiting(loggerClassName, "SQLServerBulkBatchInsertRecord");
+    }
+
+    /**
+     * Releases any resources associated with the batch.
+     * 
+     * @throws SQLServerException
+     *             when an error occurs
+     */
+    public void close() throws SQLServerException {
+    }
+
+    public DateTimeFormatter getColumnDateTimeFormatter(int column) {
+        return columnMetadata.get(column).dateTimeFormatter;
+    }
+
+    @Override
+    public Set<Integer> getColumnOrdinals() {
+        return columnMetadata.keySet();
+    }
+
+    @Override
+    public String getColumnName(int column) {
+        return columnMetadata.get(column).columnName;
+    }
+
+    @Override
+    public int getColumnType(int column) {
+        return columnMetadata.get(column).columnType;
+    }
+
+    @Override
+    public int getPrecision(int column) {
+        return columnMetadata.get(column).precision;
+    }
+
+    @Override
+    public int getScale(int column) {
+        return columnMetadata.get(column).scale;
+    }
+
+    @Override
+    public boolean isAutoIncrement(int column) {
+        return false;
+    }
+
+    private Object convertValue(ColumnMetadata cm,
+            Object data) throws SQLServerException {
+        switch (cm.columnType) {
+            /*
+             * Both BCP and BULK INSERT considers double quotes as part of the data and throws error if any data (say "10") is to be inserted into an
+             * numeric column. Our implementation does the same.
+             */
+            case Types.INTEGER: {
```

Comment:
https://github.com/Microsoft/mssql-jdbc/issues/531
https://github.com/Microsoft/mssql-jdbc/pull/538/

Do we have to convert everything to String?

Reply:
I don't think we need to. I'm following the implementation of bulk copy from SQLServerBulkCSVFileRecord. The PRs you linked are not related with these files - we can look into optimizing the bulk copy process in another PR.

Reply:
It turns out we also need to do this, actually. The alternative of using instanceof would be messy here, since the user can provide any type of numeric value through setObject (they could do 123.45f or 12345L to instantiate those numbers as float, double or long, etc), and we would need an if statement for each of those types.

That being said, I found an issue with the existing implementation - it was actually not flooring the data. The number 123.56 would be inserted as 124 into the column, whereas it should've been 123. I've fixed this issue.